File: training-data/elixir/longest_palindromic_substring.exs
defmodule LongestPalindromicSubstring do
  def find(s) do
    chars = String.to_charlist(s)
    len = length(chars)

    if len == 0 do
      ""
    else
      {start, finish} =
        0..(len - 1)
        |> Enum.reduce({0, 0}, fn i, {best_start, best_end} ->
          {s1, e1} = expand(chars, len, i, i)
          {s2, e2} = expand(chars, len, i, i + 1)

          [{best_start, best_end}, {s1, e1}, {s2, e2}]
          |> Enum.max_by(fn {s, e} -> e - s end)
        end)

      chars
      |> Enum.slice(start, finish - start + 1)
      |> List.to_string()
    end
  end

  defp expand(chars, len, left, right) do
    if left >= 0 and right < len and Enum.at(chars, left) == Enum.at(chars, right) do
      expand(chars, len, left - 1, right + 1)
    else
      {left + 1, right - 1}
    end
  end
end

IO.inspect(LongestPalindromicSubstring.find("babad"))
IO.inspect(LongestPalindromicSubstring.find("cbbd"))
IO.inspect(LongestPalindromicSubstring.find("a"))

File: training-data/matlab/longest_palindromic_substring.m
function result = longest_palindromic_substring(s)
    if isempty(s)
        result = '';
        return;
    end

    best_start = 1;
    best_end = 1;
    n = length(s);

    for i = 1:n
        [l1, r1] = expand_around_center(s, i, i);
        if (r1 - l1) > (best_end - best_start)
            best_start = l1;
            best_end = r1;
        end

        [l2, r2] = expand_around_center(s, i, i + 1);
        if (r2 - l2) > (best_end - best_start)
            best_start = l2;
            best_end = r2;
        end
    end

    result = s(best_start:best_end);
end

function [left, right] = expand_around_center(s, left, right)
    n = length(s);
    while left >= 1 && right <= n && s(left) == s(right)
        left = left - 1;
        right = right + 1;
    end
    left = left + 1;
    right = right - 1;
end

disp(longest_palindromic_substring('babad'))
disp(longest_palindromic_substring('cbbd'))
